Make ScriptPubKeyMan an actual interface and the wallet to have multiple

achow101 opened the pull request:
Continuation of wallet boxes project.

Actually makes ScriptPubKeyMan an interface which LegacyScriptPubkeyMan. Moves around functions and things from CWallet into LegacyScriptPubKeyMan so that they are actually separate things without circular dependencies.

***

Introducing the `ScriptPubKeyMan` (short for ScriptPubKeyManager) for managing scriptPubKeys and their associated scripts and keys. This functionality is moved over from `CWallet`. Instead, `CWallet` will have a pointer to a `ScriptPubKeyMan` for every possible address type, internal and external. It will fetch the correct `ScriptPubKeyMan` as necessary. When fetching new addresses, it chooses the `ScriptPubKeyMan` based on address type and whether it is change. For signing, it takes the script and asks each `ScriptPubKeyMan` for whether that `ScriptPubKeyMan` considers that script `IsMine`, whether it has that script, or whether it is able to produce a signature for it. If so, the `ScriptPubKeyMan` will provide a `SigningProvider` to the caller which will use that in order to sign.

There is currently one `ScriptPubKeyMan` - the `LegacyScriptPubKeyMan`. Each `CWallet` will have only one `LegacyScriptPubKeyMan` with the pointers for all of the address types and change pointing to this `LegacyScriptPubKeyMan`. It is created when the wallet is loaded and all keys and metadata are loaded into it instead of `CWallet`. The `LegacyScriptPubKeyMan` is primarily made up of all of the key and script management that used to be in `CWallet`. For convenience, `CWallet` has a `GetLegacyScriptPubKeyMan` which will return the `LegacyScriptPubKeyMan` or a `nullptr` if it does not have one (not yet implemented, but callers will check for the `nullptr`). For purposes of signing, `LegacyScriptPubKeyMan`'s `GetSigningProvider` will return itself rather than a separate `SigningProvider`. This will be different for future `ScriptPubKeyMan`s.

The `LegacyScriptPubKeyMan` will also handle the importing and exporting of keys and scripts instead of `CWallet`. As such, a number of RPCs have been limited to work only if a `LegacyScriptPubKeyMan` can be retrieved from the wallet. These RPCs are `sethdseed`, `addmultisigaddress`, `importaddress`, `importprivkey`, `importpubkey`, `importmulti`, `dumpprivkey`, and `dumpwallet`. Other RPCs which relied on the wallet for scripts and keys have been modified in order to take the `SigningProvider` retrieved from the `ScriptPubKeyMan` for a given script.

Overall, these changes should not effect how everything actually works and the user should experience no difference between having this change and not having it. As such, no functional tests were changed, and the only unit tests changed were those that were directly accessing `CWallet` functions that have been removed.

This PR is the last step in the [Wallet Structure Changes](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Class-Structure-Changes).

achow101 committed:
List output types in an array in order to be iterated over

achow101 committed:
Store p2sh scripts in AddAndGetDestinationForScript

achow101 committed:
Always try to sign for all pubkeys in multisig

achow101 committed:
Cleanup: Drop unused GUI learnRelatedScripts method

This commit does not change behavior.

achow101 committed:
Refactor: Replace SigningProvider pointers with unique_ptrs

Needed for future ScriptPubKeyMans which may need to create
SigningProviders dynamically and thus a normal pointer is not enough

This commit does not change behavior.

achow101 committed:
HD Split: Avoid redundant upgrades

This avoids repeaded upgrades when support for more multiple keyman references
is added in the next commit:
https://github.com/bitcoin/bitcoin/pull/16341#discussion_r322370108

achow101 committed:
Box the wallet: Add multiple keyman maps and loops

Add wallet logic for dealing with multiple ScriptPubKeyMan instances. This
doesn't change current behavior because there is still only a single
LegacyScriptPubKeyMan. But in the future the new logic will be used to support
descriptor wallets.

achow101 committed:
Refactor: Copy CWallet signals and print function to LegacyScriptPubKeyMan

This commit does not change behavior.

achow101 committed:
Refactor: Allow LegacyScriptPubKeyMan to be null

In CWallet::LoadWallet, use this to detect and empty wallet with no keys

This commit does not change behavior.

achow101 committed:
Locking: Lock cs_KeyStore instead of cs_wallet in legacy keyman

This commit only affects locking behavior and doesn't have other changes.

achow101 commented:
The final diff is very similar to #16341 just with some things dropped as they were deemed unnecessary.

Sjors commented:
For comparison, my last [ACK](https://github.com/bitcoin/bitcoin/pull/16341#issuecomment-546274939) on the original was for c37be15. That code was identical to @ryanofsky's [branch](https://github.com/ryanofsky/bitcoin/commits/pr/keyman) at  299296e51f6ea0e416aa84d7fc139195be96a58d, just using a different set of commits to get there.   This is rebased on master @25d7e2e78137d07eb612c44d19b0d496050c947a, with the last two commits (952ba99a447b619e353816cac429f825cb3bcaff & 299296e51f6ea0e416aa84d7fc139195be96a58d) dropped.  The result is f47ffe0a88111a63e1d7d98c25fbe2184215dbbe here (minus two linter changes).

ryanofsky commented:
Suggestion: there's a cluster of fairly trivial refactoring commits in this PR that I think could be pulled out into a separate PR and merged pretty quickly. Doing this would reduce the number of commits here by almost:

- [ ] b318219c51e316cfb75639f0820cc1b8017bfdfd MOVEONLY: Reorder LegacyScriptPubKeyMan methods (4/36)
- [ ] 8c7eac4de3e94adcf3ec8a0358c410a198b8b844 Refactor: Declare LegacyScriptPubKeyMan methods as virtual (5/36)
- [ ] f0339ba0b3a9d54f04b0e5f0ec80a9f791567caa Refactor: Add new ScriptPubKeyMan virtual methods (6/36)
- [ ] 5321cf7eb565834a7e69331d690543375890d516 Refactor: Move SetAddressBook call out of LegacyScriptPubKeyMan::GetNewDestination (7/36)
- [ ] 44f524e6c2116a4c57e9e47c4b130efe751cf5a7 Refactor: Move SetWalletFlag out of LegacyScriptPubKeyMan::UpgradeKeyMetadata (8/36)
- [ ] 2b92e87104868e11be9a0e2129d1340677ebc81c Refactor: Remove UnsetWalletFlag call from LegacyScriptPubKeyMan::SetHDSeed (9/36)
- [ ] 802dfb95abf9465efe34d16ece4f7e6dc8349177 Refactor: Move SetAddressBookWithDB call out of LegacyScriptPubKeyMan::ImportScriptPubKeys (10/36)
- [ ] d8d2d31e55a53bfeb3860f278e47692970bdcd84 Refactor: Move LoadKey LegacyScriptPubKeyMan method definition (11/36)
- [ ] 9f20667b5bb4f7939efa2cf16e3f12abf34ebbe5 Refactor: Move GetMetadata code out of getaddressinfo (12/36)
- [ ] 309ee44cc663c8d753729d94100bbbe694c2e091 Refactor: Move MarkUnusedAddresses code out of CWallet::AddToWalletIfInvolvingMe (13/36)
- [ ] 1ffce090c62ca340e09ace89d859843b995b0ee9 Refactor: Move Upgrade code out of CWallet::CreateWalletFromFile (14/36)
- [ ] f09be4ff1cc00390fc50147c63260983a2d98d71 Refactor: Move HavePrivateKeys code out of CWallet::CreateWalletFromFile (15/36)
- [ ] 25d904ce890f2989cbd3f677972be59d1381e98c Refactor: Move SetupGeneration code out of CWallet (16/36)
- [ ] a207dbc464a647342c55cfd3c81e9866c063cdce Refactor: Move RewriteDB code out of CWallet (17/36)
- [ ] 1cca908e4d1f855e961eb513ff87deda6a615323 Refactor: Move GetKeypoolSize code out of CWallet (18/36)
- [ ] 23008f4c36d3bd982dcdb70698b6dc05b673c4e0 Refactor: Move nTimeFirstKey accesses out of CWallet (19/36)

@achow101, would you want to pull out the commits above into a separate PR? I'd also be happy to make the PR if it would be more convenient.

achow101 commented:
@ryanofsky Done as #17304

MarcoFalke commented:
>  fanquake added this to Blockers in High-priority for review 5 hours ago 

Why?

fanquake commented:
Achow requested it via IRC. He didn't have anything in high-prio.

On Tue, 29 Oct 2019 at 16:15, MarcoFalke <notifications@github.com> wrote:

> fanquake added this to Blockers in High-priority for review 5 hours ago
>
> Why?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/17261?email_source=notifications&email_token=AAGS34SKBDHCRX7QCH2LWVLQRCKWDA5CNFSM4JFL4BW2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECR5VXY#issuecomment-547609311>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGS34RIQ5JLGSG4BKV3SCTQRCKWDANCNFSM4JFL4BWQ>
> .
>

MarcoFalke commented:
https://github.com/bitcoin/bitcoin/pull/17260#event-2751955233 was in for this week, but fair enough

achow101 committed:
Make UpgradeKeyMetadata work only on LegacyScriptPubKeyMan

ryanofsky commented:
In case its useful, rebased branch [`pr/keyman`](https://github.com/ryanofsky/bitcoin/commits/pr/keyman) 9e776722b902d8f6d095d06053b12746065baa46 -> 6ff0b88d5a98465eb79a02ceb7a46afb31d43708 ([`pr/keyman.9`](https://github.com/ryanofsky/bitcoin/commits/pr/keyman.9) -> [`pr/keyman.10`](https://github.com/ryanofsky/bitcoin/commits/pr/keyman.10)) after #17300 and #17304

ryanofsky commented:
Updated branch [`pr/keyman`](https://github.com/ryanofsky/bitcoin/commits/pr/keyman) 6ff0b88d5a98465eb79a02ceb7a46afb31d43708 -> 72df6f1050d27a13435fab77f20c1a0e08fdbf51 ([`pr/keyman.10`](https://github.com/ryanofsky/bitcoin/commits/pr/keyman.10) -> [`pr/keyman.11`](https://github.com/ryanofsky/bitcoin/commits/pr/keyman.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/keyman.10..pr/keyman.11)) to fix exception in `CWallet::GetLegacyScriptPubKeyMan()` caused by being called from `CWallet::UpgradeKeyMetadata()` early in initialization. This makes tests pass.

achow101 commented:
Rebased

ryanofsky commented:
Now that most of code movement is done, to make more progress on this, I'd say the following individual commits are complicated enough to be their own PRs which could all be reviewed and merged in parallel:

- [ ] b5f456eb91e687ed39bc38e5b9b4e5aca0434515 Always try to sign for all pubkeys in multisig (2/20)
- [ ] a1297615cce13bba3c16412418e0522e3f7f9eff Store p2sh scripts in AddAndGetDestinationForScript (3/20)
- [ ] 09d99fc720998c892b60c26ff02e7dc11f69fee4 Refactor: Move encryption code between KeyMan and Wallet (4/20)
- [ ] 11157cffc8925fe129d045d0238106db3bfb444a Refactor: Allow LegacyScriptPubKeyMan to be null (6/20)
- [ ] a0b524c10f6944195f8fffa2a6558e002647fe23 Refactor: Require scriptPubKey to get wallet SigningProvider (7/20)
- [ ] ac8cbde93df8cf365a08f472d2018f9667fae7c9 IsMine: Set state to WATCH_ONLY if we can get the pubkey (14/20)

Then there is a set of keypool tweaks and fixes, which could go in a single PR (reviewed in parallel with PRs above):

- [ ] 71f1a7ff10408211b7f86ee3c932a3d264441643 Key pool: Move CanGetAddresses call (8/20)
- [ ] 5733cc0303673d58e887313965e5fb045f08e472 Key pool: Move LearnRelated and GetDestination calls (9/20)
- [ ] 922252b6ff74a9d0da60c7bd87217eb5b938e220 Key pool: Move TopUp call (10/20)
- [ ] 4fb82dcb7fac2f3684b9c4345946956488066d7a Key pool: Change ReturnDestination interface to take address instead of key (11/20)
- [ ] 0c823e4d5c687fe116d2de550888558a3005b184 Key pool: Make TopUp fail if unexpected wallet flags are set (12/20)
- [ ] 68a35f25bf3036d3390243c1a06887681d144586 Key pool: Fix omitted pre-split count in GetKeyPoolSize (13/20)

After the encryption change above is in (09d99fc720998c892b60c26ff02e7dc11f69fee4), the cs_wallet/cs_keystore locking change could be a quick followup PR:

- [ ] 558e0bb42cc0336bde941fb3a6e871e98a96962f Locking: Lock cs_KeyStore instead of cs_main in legacy keyman (5/20)

After that, the remaining commits in this PR would add the new key manager maps and lookup code, and do final cleanups and tweaks:

- [ ] fbcf6ad0ee6a87ac718c3ee2d9fa7663e18495a2 List output types in an array in order to be iterated over (1/20)
- [ ] 85a1bb8e3a363eeca8f15505b716d6c04733cd11 HD Split: Avoid redundant upgrades (15/20)
- [ ] 2fa9e854da0641d289b5d975ab3d15449fe59deb Box the wallet: Add multiple keyman maps and loops (16/20)
- [ ] dd447f855324d5c6cebc2054db03d2fb5fdc4149 Refactor: Copy CWallet signals and print function to LegacyScriptPubKeyMan (17/20)
- [ ] ddf78847a8abc62b3339fe98aea467a2fce894c6 Cleanup: Update message strings and comments (18/20)
- [ ] 32ec05d7aec6574928f5ca019c61ce24dccb1c25 Cleanup: Drop unused GUI learnRelatedScripts method (19/20)
- [ ] 72df6f1050d27a13435fab77f20c1a0e08fdbf51 Refactor: Replace SigningProvider pointers with unique_ptrs (20/20)

ryanofsky commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -292,8 +292,7 @@ class LegacyScriptPubKeyMan : public ScriptPubKeyMan, public FillableSigningProv
 
     void MarkUnusedAddresses(const CScript& script) override;
 
-    //! Upgrade stored CKeyMetadata objects to store key origin info as KeyOriginInfo
```
In commit "Locking: Lock cs_KeyStore instead of cs_main in legacy keyman" (90287119f4bab55e3ffb8e5b9139d19f11eebd61):

Looks like comment this line was unintentionally deleted (mistake during rebase)

achow101 replied:
Added it back in.

ryanofsky reviewed (commented):
Current rebase d03af791a8b7a290c0d883ecf6642ead3ffe6a6a looks basically same as the rebase I did, but I think it avoids a bug in CWallet::UpgradeKeyMetadata where my rebase would incorrectly set the WALLET_FLAG_KEY_ORIGIN_METADATA if legacy keyman was null (tests passed anyway so I guess coverage is not great here). The cleanup change to group together all wallet flag set/unset methods in wallet.h was also dropped.

achow101 commented:
"Refactor: Move encryption code between KeyMan and Wallet" was split into #17369

"Refactor: Require scriptPubKey to get wallet SigningProvider" was split into #17371 

The Keypool commits were split into #17373

"IsMine: Set state to WATCH_ONLY if we can get the pubkey" was split into #17374

I had some issues with making "Refactor: Allow LegacyScriptPubKeyMan to be null" standalone, there were some segfaults and other conflicting changes since it comes after the locking and encryption commits. So I don't think I will make this standalone.

Both "Always try to sign for all pubkeys in multisig" and "Store p2sh scripts in AddAndGetDestinationForScript" aren't needed until "Box the wallet: Add multiple keyman maps and loops". I don't think they make sense by themselves, and they aren't really complicated. So I would prefer to bundle them with "Box the wallet: Add multiple keyman maps and loops".

I'll PR the locking commit after the encryption commit is merged.

promag commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -278,6 +278,8 @@ bool LegacyScriptPubKeyMan::GetReservedDestination(const OutputType type, bool i
     if (!ReserveKeyFromKeyPool(index, keypool, internal)) {
         return false;
     }
+    LearnRelatedScripts(keypool.vchPubKey, type);
```
3499674136167f662e5b0ef0b599c02f5a959558

You gave ACK in #17237 so I wonder I is this here?

ryanofsky replied:
> [3499674](https://github.com/bitcoin/bitcoin/commit/3499674136167f662e5b0ef0b599c02f5a959558)
> 
> You gave ACK in #17237 so I wonder why is this here?

Better to ask this question in #17373 where this commit was moved than this PR. Specifically [17373/commits/35c37b5](https://github.com/bitcoin/bitcoin/pull/17373/commits/35c37b59c5bb918c15aca3092b6506ef3974e50c)

Probably this code change just precedes your PR

ryanofsky replied:
It would be nice to get more review on #17373, even if this question won't be raised. The changes there are confusing to me, and might be more straightforward to you. (The changes there are minor and seem safe but I don't understand why they are made.)

promag replied:
You mean #17373, not #17237 right? If so I'll review the first.

ryanofsky commented:
Maybe say at the top of the PR description that this builds on #17369, #17371, #17373, and those PRs should be reviewed first to avoid misplaced review comments here.

Can also link to the descriptor wallet project url which lists all PRs https://github.com/bitcoin/bitcoin/projects/12

Sjors commented:
For those interested in helping out, please review #17369, #17369, #17537 and #17261.

ryanofsky commented:
> For those interested in helping out, please review #17369, #17537 and #17261.

Would add #17373. You can also check https://github.com/bitcoin/bitcoin/projects/12 for the latest list of descriptor wallet pull requests.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3297,21 +3297,21 @@ UniValue signrawtransactionwithwallet(const JSONRPCRequest& request)
         }
     }
     if (providers.size() == 0) {
         // When there are no available providers, use DUMMY_SIGNING_PROVIDER so we can check if the tx is complete
```
comment is now out of date

achow101 replied:
Updated

achow101 commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -196,8 +196,16 @@ class ScriptPubKeyMan
+    virtual const SigningProvider* GetSigningProvider(const CScript& script) const { return nullptr; }
+
+    /** Whether this ScriptPubKeyMan can provide a SigningProvider (via GetSigningProvider) that, combined with
+      * sigdata, can produce a valid signature.
```
Is that not stated by the "combined with sigdata" part?

instagibbs replied:
If I think of better wording I'll let you know.

achow101 committed:
refactor: define a UINT256_ONE global constant

Instead of having a uint256 representations of one scattered throughout
where it is used, define it globally in uint256.h

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3073,7 +3076,7 @@ size_t CWallet::KeypoolCountExternalKeys()
     AssertLockHeld(cs_wallet);
```
No, they still apply to other SPKMs.

achow101 commented on src/qt/test/wallettests.cpp:
```diff
@@ -141,8 +141,7 @@ void TestGUI(interfaces::Node& node)
         auto locked_chain = wallet->chain().lock();
-        LOCK(wallet->cs_wallet);
-        AssertLockHeld(spk_man->cs_wallet);
+        LOCK2(wallet->cs_wallet, spk_man->cs_KeyStore);
```
Fixed.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -491,6 +519,11 @@ const CKeyMetadata* LegacyScriptPubKeyMan::GetMetadata(const CTxDestination& des
 
+uint256 LegacyScriptPubKeyMan::GetID() const
+{
+    return uint256S("0000000000000000000000000000000000000000000000000000000000000001");
```
Since this value of one is used in a couple of other places,  I've added a commit refactor those uses and adding a constant `UINT256_ONE` to uint256.h. So we use that constant and so do the other places that use that value of one.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -4056,24 +4061,81 @@ bool CWallet::Unlock(const CKeyingMaterial& vMasterKeyIn, bool accept_no_keys)
+
+ScriptPubKeyMan* CWallet::GetScriptPubKeyMan(const OutputType& type, bool internal) const
+{
+    if (internal) {
```
Done

achow101 commented on src/wallet/wallet.h:
```diff
@@ -1116,13 +1123,22 @@ class CWallet final : public WalletStorage, private interfaces::Chain::Notificat
         LogPrintf(("%s " + fmt).c_str(), GetDisplayName(), parameters...);
     };
 
+    //! De-duplicates and returns all ScriptPubKeyMans in m_internal_spk_managers and m_external_spk_managers
```
Done

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/17261/commits/db4a741d30a44ebd975c905cef0f9578c816e8f2

updates are as requested

Sjors commented:
85901a994ea00a6fa9af44eeaed418b968829f26 `Refactor: Allow LegacyScriptPubKeyMan to be null` breaks `sethdseed` for blank wallets, because of the `EnsureLegacyScriptPubKeyMan` check. None of the tests catch this.

I'm also seeing `"keypoololdest": 1578118989` (instead of `0`) for blank wallets, not when created, but after a restart. That may be a pre-existing issue.

Sjors reviewed (commented):
Other than the above `sethdseed ` issue: code review ACK db4a741d30a44ebd975c905cef0f9578c816e8f2

Suggestion for followup: add tests for the PSBT related change in dec753da5045be56623ec29155dc4ce5e416e1a5, because I can drop it without breaking any wallet functional test.

instagibbs commented:
> I'm also seeing "keypoololdest": 1578118989 (instead of 0) for blank wallets, not when created, but after a restart. That may be a pre-existing issue.

From my reading of the code that's what it's "supposed" to do when the keypool is empty on master. Help should probably detail that.

kallewoof commented on src/uint256.h:
```diff
@@ -144,4 +144,6 @@ inline uint256 uint256S(const std::string& str)
     return rv;
 }
 
+uint256& UINT256_ONE();
```
Why not simply make this `extern uint256 UINT256_ONE;` instead of a singleton?

achow101 replied:
Clang didn't like it. Something about static initialization order resulting in a potential uninitialized object access.

kallewoof replied:
Weird. :/ I've seen that once before too. Sounds like a compiler bug tbh.

kallewoof commented:
No idea why github decided to post my review, I'm only halfway done.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1876,7 +1876,7 @@ static UniValue keypoolrefill(const JSONRPCRequest& request)
 
     if (pwallet->GetKeyPoolSize() < kpSize) {
-        throw JSONRPCError(RPC_WALLET_ERROR, "Error refreshing keypool.");
+        throw JSONRPCError(RPC_WALLET_ERROR, strprintf("Error refreshing keypool. %d", pwallet->GetKeyPoolSize()));
```
This looks like debug leftovers.

achow101 replied:
Removed

kallewoof commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -442,4 +442,20 @@ class LegacyScriptPubKeyMan : public ScriptPubKeyMan, public FillableSigningProv
+private:
+    const LegacyScriptPubKeyMan& spk_man;
+public:
+    LegacySigningProvider(const LegacyScriptPubKeyMan& spk_man) : spk_man(spk_man) {}
```
Call it `m_spk_man` and avoid `a(a)` → `m_spk_man(spk_man)` ?

achow101 replied:
Done

kallewoof reviewed (commented):
Code review ACK

kallewoof commented:
@instagibbs 
> From my reading of the code that's what it's "supposed" to do when the keypool is empty on master; it will just use current clock time. Help should probably detail that.

I think the problem here is that it is `0` before restart, not that it uses current clock time.

kallewoof commented:
Tested ACK db4a741d30a44ebd975c905cef0f9578c816e8f2

Tested on top of signet patch. Received and sent funds. All seems to work a-OK.

achow101 commented on src/outputtype.h:
```diff
@@ -47,4 +50,3 @@ std::vector<CTxDestination> GetAllDestinationsForKey(const CPubKey& key);
 CTxDestination AddAndGetDestinationForScript(FillableSigningProvider& keystore, const CScript& script, OutputType);
 
 #endif // BITCOIN_OUTPUTTYPE_H
-
```
There's an extra newline that's being removed.

kallewoof replied:
Oh, I misread then. +1 removing extra NL.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -572,9 +573,9 @@ bool CWallet::EncryptWallet(const SecureString& strWalletPassphrase)
-                if (!spk_man->SetupGeneration(true)) {
+        for (const auto& spk_man_pair : m_spk_managers) {
+            if (spk_man_pair.second->IsHDEnabled()) {
+                if (!spk_man_pair.second->SetupGeneration(true)) {
```
Done

achow101 commented on src/outputtype.cpp:
```diff
@@ -80,22 +80,30 @@ CTxDestination AddAndGetDestinationForScript(FillableSigningProvider& keystore,
+    case OutputType::LEGACY: {
+        keystore.AddCScript(GetScriptForDestination(sh));
+        return sh;
+    }
```
They were needed in a previous iteration. Removed.

achow101 commented on src/outputtype.cpp:
```diff
@@ -80,22 +80,30 @@ CTxDestination AddAndGetDestinationForScript(FillableSigningProvider& keystore,
         // Check if the resulting program is solvable (i.e. doesn't use an uncompressed key)
-        if (!IsSolvable(keystore, witprog)) return ScriptHash(script);
+        if (!IsSolvable(keystore, witprog)) {
+            keystore.AddCScript(GetScriptForDestination(sh));
```
Done

achow101 commented:
Fixed the sethdseed problem.

Sjors commented:
Code review re-ACK b5afe999206ee6982fd555b490fae6dd50a1c0c3

I assume we don't want to  add a `LegacyScriptPubKeyMan` to a (future) descriptor based wallet. That's working correctly I believe, because  `EnsureLegacyScriptPubKeyMan(*wallet, true)` calls `SetupLegacyScriptPubKeyMan` which fails if any other `ScriptPubKeyMan` types exist. But I would prefer if `EnsureLegacyScriptPubKeyMan` checked this directly.

kallewoof commented:
Is b5afe999206ee6982fd555b490fae6dd50a1c0c3 not meant to be squashed?

promag committed:
wallet: Improve CWallet:MarkDestinationsDirty

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -569,7 +574,7 @@ bool CWallet::EncryptWallet(const SecureString& strWalletPassphrase)
 
         // if we are using HD, replace the HD seed with a new one
-        if (auto spk_man = m_spk_man.get()) {
+        if (auto spk_man = GetLegacyScriptPubKeyMan()) {
```
Are we not going to replace the seed for descriptor wallets?

achow101 replied:
No, descriptor wallets have a different way of doing it.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/17261#discussion_r363792482

> No, descriptor wallets have a different way of doing it.

In commit "Box the wallet: Add multiple keyman maps and loops" (c729afd0a3b74a3943e4c359270beaf3e6ff8a7b):

It seems bad to do nothing at all and keep using unencrypted seeds for non-legacy wallets, and the comment here doesn't here doesn't explain any legacy/nonlegacy distinction. It would seem safer to trigger some kind of error for non-legacy wallets if planned behavior hasn't been implemented yet, than to just do nothing.

If there is supposed to be legacy / non-legacy distinction, it would seem better to move legacy code out of CWallet into LegacyScriptPubKeyMan in a `PostEncrypt` or similar ScriptPubKeyMan virtual method

instagibbs commented:
> EnsureLegacyScriptPubKeyMan(*wallet, true) calls SetupLegacyScriptPubKeyMan which fails if any other ScriptPubKeyMan types exist

It'll hit asserts, right? Yeah I'd rather things be explicitly checked and handled instead if possible.

achow101 commented:
> I assume we don't want to add a `LegacyScriptPubKeyMan` to a (future) descriptor based wallet. That's working correctly I believe, because `EnsureLegacyScriptPubKeyMan(*wallet, true)` calls `SetupLegacyScriptPubKeyMan` which fails if any other `ScriptPubKeyMan` types exist. But I would prefer if `EnsureLegacyScriptPubKeyMan` checked this directly.

What do you mean by "checked this directly"? Have the same wallet flag check or something else?

> It'll hit asserts, right? Yeah I'd rather things be explicitly checked and handled instead if possible.

The descriptor wallet PR checks the wallet flag and will return early if the descriptor wallet flag is set. No asserts.

> Is [b5afe99](https://github.com/bitcoin/bitcoin/commit/b5afe999206ee6982fd555b490fae6dd50a1c0c3) not meant to be squashed into [d8a3c78](https://github.com/bitcoin/bitcoin/commit/d8a3c7830e1292c8b7d52b5a02fdd1d2459de3ef)?

I guess it can be. Hadn't really noticed that.

Sjors commented:
> The descriptor wallet PR checks the wallet flag and will return early if the descriptor wallet flag is set. No asserts.

It's good that `EnsureLegacyScriptPubKeyMan` never gets called for a descriptor wallet, but it would be better if it explicitly checked that there is no other `ScriptPubKeyMan` out there. Currently it only checks that there's no `LegacyScriptPubKeyMan` out there, but then hit an assert later on if it wasn't the only `ScriptPubKeyMan`.

achow101 commented:
> > The descriptor wallet PR checks the wallet flag and will return early if the descriptor wallet flag is set. No asserts.
> 
> It's good that `EnsureLegacyScriptPubKeyMan` never gets called for a descriptor wallet, but it would be better if it explicitly checked that there is no other `ScriptPubKeyMan` out there. Currently it only checks that there's no `LegacyScriptPubKeyMan` out there, but then hit an assert later on if it wasn't the only `ScriptPubKeyMan`.

I don't quite understand what you want me to do here. We don't have any other `ScriptPubKeyMan`s yet, so we can't check if it is something else. And if there are other `ScriptPubKeyMan`s in a wallet that has a `LegacyScriptPubKeyMan`, then it definitely should assert and crash because that's not a safe or consistent state for the wallet to be in.

Sjors commented:
Instead of:
```cpp
 LegacyScriptPubKeyMan* spk_man = nullptr;
 if (also_create) {
     spk_man = wallet.GetOrCreateLegacyScriptPubKeyMan();
 } else {
     spk_man = wallet.GetLegacyScriptPubKeyMan();
 }
```

This seems more clear:
```cpp
 LegacyScriptPubKeyMan* spk_man = wallet.GetLegacyScriptPubKeyMan();
 if (!spk_man && also_create) {
     spk_man = wallet.GetOrCreateLegacyScriptPubKeyMan();
 }
```

And instead of:
```cpp
void CWallet::SetupLegacyScriptPubKeyMan()
{
    if (m_internal_spk_managers.empty() && m_external_spk_managers.empty() && m_spk_managers.empty()) {
        auto spk_manager = std::unique_ptr<ScriptPubKeyMan>(new LegacyScriptPubKeyMan(*this));
```

Try:
```cpp
void CWallet::SetupLegacyScriptPubKeyMan()
{
    if(m_internal_spk_managers.count(OutputType::LEGACY) > 0) {
        return;
    }
    assert(m_internal_spk_managers.empty() && m_external_spk_managers.empty() && m_spk_managers.empty());
    auto spk_manager = std::unique_ptr<ScriptPubKeyMan>(new LegacyScriptPubKeyMan(*this));
```

(this top `if` statement might be avoidable by having a separate `CreateLegacyScriptPubKeyMan()`)

achow101 commented:
Rebased.

@Sjors Done. I changed your suggestion for `SetupLegacyScriptPubKeyMan` to check whether any of the `*spk_managers` maps has a `ScrriptPubKeyMan`.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -124,9 +124,12 @@ void EnsureWalletIsUnlocked(const CWallet* pwallet)
 }
 
-LegacyScriptPubKeyMan& EnsureLegacyScriptPubKeyMan(CWallet& wallet)
+LegacyScriptPubKeyMan& EnsureLegacyScriptPubKeyMan(CWallet& wallet, bool also_create)
```
What's the general usage for this new boolean? I see it's set to true for most importing, but not for `addmultisigaddress` for instance. A guiding comment on usage would be helpful.

achow101 replied:
Added a comment.

It's to be used for RPCs that make a blank wallet no longer blank. `addmultisigaddress` wasn't doing this before, although I suppose the expected behavior of it with a blank wallet isn't really known (i.e. what do we want it to do?).

instagibbs replied:
Hm, yeah it would be a weird non-blank wallet. I'm fine as-is for now.

instagibbs commented on src/outputtype.cpp:
```diff
@@ -80,22 +80,30 @@ CTxDestination AddAndGetDestinationForScript(FillableSigningProvider& keystore,
             return witdest;
         } else {
-            return ScriptHash(witprog);
+            ScriptHash wsh = ScriptHash(witprog);
```
micro-nit: wsh is a bit of a minsnomer, it's `sh_w`, or something

achow101 replied:
Done

instagibbs commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -384,7 +384,7 @@ bool LegacyScriptPubKeyMan::Upgrade(int prev_version, std::string& error)
     }
     // Upgrade to HD chain split if necessary
-    if (m_storage.CanSupportFeature(FEATURE_HD_SPLIT)) {
+    if (m_storage.CanSupportFeature(FEATURE_HD_SPLIT) && hdChain.nVersion < 2 /* VERSION_HD_CHAIN_SPLIT */) {
```
any way to actually use the constant rather than referring to it and hoping the value never changes?

achow101 replied:
Yes. the proper header is being imported now, so no need for this hack. Changed it.

Also, the value should never change because its persisted to wallet files and changing it would break backwards compatibility.

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/17261/commits/bbf50a006410bd3e56908983664c5a1a0b475c2c

Sjors commented:
Code review ACK bbf50a006410bd3e56908983664c5a1a0b475c2c

achow101 commented:
Rebased

Sjors commented:
re-ACK c985aea9be828c0a49eb81f5a5d54b42147ef20c; just a rebase with merged #17843 content dropped.

meshcollider commented on src/wallet/wallet.cpp:
```diff
@@ -4135,7 +4188,26 @@ LegacyScriptPubKeyMan* CWallet::GetOrCreateLegacyScriptPubKeyMan()
+        return;
+    }
+
+    assert(m_internal_spk_managers.empty() && m_external_spk_managers.empty() && m_spk_managers.empty());
```
This assert seems a little unnecessary considering the if-statement above

achow101 replied:
Removed

meshcollider commented on src/wallet/wallet.cpp:
```diff
@@ -4135,7 +4188,26 @@ LegacyScriptPubKeyMan* CWallet::GetOrCreateLegacyScriptPubKeyMan()
+        assert(m_internal_spk_managers.at(type) == spk_man);
+        assert(m_external_spk_managers.at(type) == spk_man);
+    }
+    assert(m_spk_managers.size() == 1);
```
In what case would these asserts ever be hit, directly following the setup code? Is there a scenario you can think of?

achow101 replied:
I don't think so. I've removed these asserts.

achow101 replied:
I think a better solution would be to have this return a bool, but that's an improvement we can make in the future as it's really a belt-and-suspenders check.

meshcollider commented on src/wallet/wallet.cpp:
```diff
@@ -3906,7 +3910,7 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
         // our wallet birthday (as adjusted for block time variability)
         Optional<int64_t> time_first_key;
-        if (auto spk_man = walletInstance->m_spk_man.get()) {
+        for (auto spk_man : walletInstance->GetActiveScriptPubKeyMans()) {
```
Why only check the "active" SPKMans here?

achow101 replied:
Changed to check all.

instagibbs replied:
not sure this was needed, but it's a safe change

meshcollider reviewed (commented):
Strong/obvious concept ACK. I'm excited to get this merged soon. I've done an initial review:

- [x] (didn't review in depth, just read through it) 0bceafb874dbbc2994de1fe10d37782ffdcb0e55 Locking: Lock cs_KeyStore instead of cs_wallet in legacy keyman
- [x] 6c953ae84f4272d036602d5c7cf19456d9f39863 Refactor: Allow LegacyScriptPubKeyMan to be null
- [x] 14f7284eb52dafa9fa36955f39296e31bf534cdd List output types in an array in order to be iterated over
- [x] 97eac193ed167ba48c2b57fd16a1b52170fc87dc Always try to sign for all pubkeys in multisig
- [x] ec0d430be8888c60e763c7c12a545674eb876c40 Store p2sh scripts in AddAndGetDestinationForScript
- [x] 9941be111049cd9463ad69ba5a4f1e9fbaa97d48 Make UpgradeKeyMetadata work only on LegacyScriptPubKeyMan
- [x] d27e13f5050b1ed24a49e351a2fa607297104794 HD Split: Avoid redundant upgrades
- [x] de37afd27b5407e742797cdb35f935f9fabfd03a refactor: define a UINT256_ONE global constant
- [x] b4c48fbd5602f57e9bc59de83c6c256fdd0f2d71 Box the wallet: Add multiple keyman maps and loops
- [x] f62eaa952f24c54dcc63353fb06a61b979b56577 Refactor: Copy CWallet signals and print function to LegacyScriptPubKeyMan
- [x] 0c4cb4956003c83706c812dfd699f5b22ab58d44 Cleanup: Update message strings and comments
- [x] 8b08ca79f19c65681a3acbcc236c6fb516169a45 Cleanup: Drop unused GUI learnRelatedScripts method
- [x] c985aea9be828c0a49eb81f5a5d54b42147ef20c Refactor: Replace SigningProvider pointers with unique_ptrs

A couple of comments inline. I'll test this out soon too.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2993,10 +2993,9 @@ DBErrors CWallet::LoadWallet(bool& fFirstRunRet)
         }
     }
 
+    // This wallet is in its first run if all of these are empty and this isn't blank or no privkeys
```
Edited the comment.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1056,7 +1055,7 @@ bool LegacyScriptPubKeyMan::NewKeyPool()
             return false;
         }
-        WalletLogPrintf("LegacyScriptPubKeyMan::NewKeyPool rewrote keypool\n");
+        WalletLogPrintf("CWallet::NewKeyPool rewrote keypool\n");
```
On further examination, that entire commit doesn't make sense, so I've dropped it.

instagibbs reviewed (approved):
re-utACK https://github.com/bitcoin/bitcoin/pull/17261/commits/126d6945e38cdc47b1227cdacfaa456714efcdf5

Sjors commented:
re-utACK 126d6945e38cdc47b1227cdacfaa456714efcdf5

laanwj commented on src/outputtype.h:
```diff
@@ -27,6 +28,8 @@ enum class OutputType {
     CHANGE_AUTO,
 };
 
+const std::array<OutputType, 3> OUTPUT_TYPES = {OutputType::LEGACY, OutputType::P2SH_SEGWIT, OutputType::BECH32};
```
Doesn't defining the value of non-primitive constants in the header file potentially cause linker issues in C++ (or duplication, as this literal data will be included in multiple files)?. I think it's better to only declare the variable here then define the value in the `cpp`.

achow101 replied:
I don't think this would cause any issues since it's const. I also haven't seen any linker warnings or errors about this. I think there would only be some duplication, but isn't that the same for any constant global variable that is defined in a header?

kallewoof replied:
If you are fine trusting the word of [Microsoft](https://docs.microsoft.com/en-us/cpp/cpp/const-cpp?view=vs-2019),

> In C, constant values default to external linkage, so they can appear only in source files. In C++, constant values default to internal linkage, which allows them to appear in header files.

Another source said C++ does not normally create storage for consts, which I assume means they work like #defines, in which case there shouldn't be wasted duplicate data entries either.

sipa replied:
If you replace the {...} initialized with a function call, and then have the constant in multiple translation units, the function will get called once for each.

achow101 replied:
I moved the initialization to the cpp file

instagibbs commented:
re-utACK https://github.com/bitcoin/bitcoin/pull/17261/commits/3f373659d732a5b1e5fdc692a45b2b8179f66bec

Sjors commented:
re-utACK 3f373659d732a5b1e5fdc692a45b2b8179f66bec (it still compiles on macOS after https://github.com/bitcoin/bitcoin/pull/17261#discussion_r370377070)

meshcollider commented:
Tested re-ACK 3f373659d732a5b1e5fdc692a45b2b8179f66bec

I think this is ready to go in

ryanofsky commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -384,7 +384,7 @@ bool LegacyScriptPubKeyMan::Upgrade(int prev_version, std::string& error)
     }
     // Upgrade to HD chain split if necessary
-    if (m_storage.CanSupportFeature(FEATURE_HD_SPLIT)) {
+    if (m_storage.CanSupportFeature(FEATURE_HD_SPLIT) && CHDChain::VERSION_HD_CHAIN_SPLIT) {
```
In commit "HD Split: Avoid redundant upgrades" (415afcccd3e5583defdb76e3a280f48e98983301):

This seems buggy and the change looks like it has no effect. Is this supposed to say `&& hdChain.nVersion < CHDChain::VERSION_HD_CHAIN_SPLIT`?

ryanofsky reviewed (commented):
Post-merge code review ACK 3f373659d732a5b1e5fdc692a45b2b8179f66bec. I did post two questions in review comments below that I think would be good to resolve.

Changes since last review (72df6f1050d27a13435fab77f20c1a0e08fdbf51):
- Dropped commit daf6d548de3efdfdb974bb52948f6086fb0e2ecc "Set state to WATCH_ONLY if we can get the pubkey"
- Added `EnsureLegacyScriptPubKeyMan` create argument to make RPCs like `sethdseed` work on empty wallets
- `coinselector_tests` setup fix
- `OUTPUT_TYPES` definition move
- `AddAndGetDestinationForScript` formatting tweaks
- Split commit 01b4511206e399981a77976deb15785d18db46ae "Make UpgradeKeyMetadata work only on LegacyScriptPubKeyMan" out of commit 264917ff4343b678007d89373aba6e99d521001c "Add multiple keyman maps and loops"
- `LegacyScriptPubKeyMan::Upgrade` hd split version check no longer added
- Replacement hd seed generation skipped in `CWallet::EncryptWallet` for nonlegacy wallets
- `CWallet::GetAllScriptPubKeyMans` method added and used by `CreateWalletFromFile` to not skip inactive keyman instances when figuring out first key time
- `CWallet::GetScriptPubKeyMan` internal true/false code deduped
- `CWallet::SetupLegacyScriptPubKeyMan` asserts removed
- `LegacySigningProvider` `m_spk_man` member rename and comment tweak

anezdoank reviewed (approved):
$ ssh_keygen_p
# start the SSH key creation proses
